iree/turbine/kernel/wave/compile.py

```diff
@@ -0,0 +1,132 @@
+from typing import Any, List, Dict
+from .._support.indexing import IndexingContext, IndexExpr
+from ..compiler import kernel_codegen, host_codegen
+from .compile_options import WaveCompileOptions
+
+from .cache import (
+    is_cache_enabled,
+    get_cache_manager,
+    WAVE_RUNTIME_DIR,
+)
+from .utils.compile_utils import compile_to_vmfb
+from .utils.general_utils import _write_file
+from iree.turbine.kernel._support.context import push, pop
+
+
+class WaveKernel:
+    """
+    Represents a wave kernel that can be invoked by the user.
+    """
+
+    def __init__(self, options: WaveCompileOptions, executable: Any, asm: str):
+        self.options = options
+        self.cached_kernel_hash = self.options.kernel_hash
+        self.subs = IndexingContext.current().subs
+        self.executable = executable
+        self.asm = asm
+
+    def __call__(self, *args, **kwargs):
+        return self.invoke(*args, **kwargs)
+
+    def invoke(self, *args, **kwargs):
+        """
+        Invokes the wave kernel with the given arguments.
+        Returns the assembly code of the compiled kernel.
+        """
+        from .utils.general_utils import invoke_vmfb
+
+        # Partition arguments into kernel inputs and outputs.
+        kernel_inputs = []
+        kernel_outputs = []
+        for arg, b in zip(args, self.config.kernel_sig.kernel_buffer_bindings):
+            usage = b.kernel_buffer_type.usage
+            if usage == kernel_codegen.KernelBufferUsage.INPUT:
+                kernel_inputs.append(arg)
+
+            if usage == kernel_codegen.KernelBufferUsage.OUTPUT:
+                kernel_outputs.append(arg)
+
+        # Get cached kernel when available.
+        cache_enabled = is_cache_enabled()
+        kernel_hash = None
+        if cache_enabled:
```

Comment:
None of the caching logic depends on invoke args. You can gen vmfb once in `__init__` (from cache or otherwise) and just use it here without any additional checks.

Reply:
good point, thanks that should speed up things!